File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
from collections import deque

K = int(input())

M, N = map(int, input().split())

board = []

"""
00 01 02 03 04
10 11 12 13 14
20 21 22 23 24
30 31 32 33 34
40 41 42 43 44

-1,-2
-2,-1
-2,1
-1,2
1,2
2,1
2,-1
1,-2
"""

nightdx = [-1,-2,-2,-1,1,2,2,1]
nightdy = [-2,-1,1,2,2,1,-1,-2]

dx = [1,0,-1,0]
dy = [0,1,0,-1]


for i in range(N):
    tmp = list(map(int, input().split()))

    board.append(tmp)

vis = [[[False for _ in range(100)] for _ in range(M)] for _ in range(N)]
cnt = [[float("INF") for _ in range(M)] for _ in range(N)]

answer = float("INF")

# for _ in range(N):
#     print()

Q = deque()
Q.append([0,0,0,0]) # x,y,jumpcnt, movecnt
vis[0][0][0] = True
while(len(Q) != 0):
    # tmp = list(Q)
    # tmp.sort(key=lambda x : x[2])
    # Q = deque(tmp)
    cur = Q.popleft()

    # print(cur)
    # print()
    # for i in range(N):
    #     print(vis[i])
    # 
    # print()

    if cur[0] == N-1 and cur[1] == M-1:
        answer = min(answer, cur[3])


    if cur[2] < K:
        for dir in range(8):
            nx = cur[0] + nightdx[dir]
            ny = cur[1] + nightdy[dir]

            if nx < 0 or nx >= N or ny < 0 or ny >= M:
                continue

            if vis[nx][ny][cur[2]+1] == True or board[nx][ny] == 1:
                continue

            Q.append([nx,ny, cur[2]+1, cur[3]+1])
            vis[nx][ny][cur[2]+1] = True

    for dir in range(4):
        nx = cur[0] + dx[dir]
        ny = cur[1] + dy[dir]

        if nx < 0 or nx >= N or ny < 0 or ny >= M:
            continue

        if vis[nx][ny] == 1:
            Q.append([nx, ny, cur[2], cur[3] + 1])
            vis[nx][ny] = 2
            continue

        if vis[nx][ny][cur[2]] == True or board[nx][ny] == 1:
            continue

        Q.append([nx, ny, cur[2], cur[3] + 1])
        vis[nx][ny][cur[2]] = True


if answer == float("INF"):
    answer = -1
print(answer)

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int K, N, M;
    static int[][] board;
    static boolean[][][] visited;
    static int answer = Integer.MAX_VALUE;

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, -1, 0, 1};

    static int[] ddx = {-2, -1, 1, 2, 2, 1, -1, -2};
    static int[] ddy = {1, 2, 2, 1, -1, -2, -2, -1};

    public static void main(String[] args) throws IOException {
        init();
        solution();


        if (answer == Integer.MAX_VALUE) {
            answer = -1;
        }

        System.out.println(answer);
    }

    private static void solution() {

        ArrayDeque<tuple> Oq = new ArrayDeque<>();
        visited = new boolean[K+1][N][M];

        Oq.add(new tuple(0, 0, 0, 0));
        visited[0][0][0] = true;

        while (!Oq.isEmpty()) {

            tuple t = Oq.removeFirst();

            if (t.x == N - 1 && t.y == M - 1) {
                answer = Math.min(answer, t.cnt);
                break;
            }


            for (int dir = 0; dir < 4; dir++) {
                int nx = t.x + dx[dir];
                int ny = t.y + dy[dir];

                if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
                if (visited[t.move][nx][ny] || board[nx][ny] == 1) continue;

                Oq.add(new tuple(t.move, nx, ny, t.cnt + 1));
                visited[t.move][nx][ny] = true;

            }


            if (t.move < K) {
                for (int dir = 0; dir < 8; dir++) {
                    int nx = t.x + ddx[dir];
                    int ny = t.y + ddy[dir];

                    if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
                    if (visited[t.move + 1][nx][ny] || board[nx][ny] == 1) continue;

//                    System.out.println("ddx" + t.move+" "+ nx + " " + ny);

                    Oq.add(new tuple(t.move + 1, nx, ny, t.cnt + 1));
                    visited[t.move + 1][nx][ny] = true;
                }
            }
        }
    }

    static void see() {

//        for (int i = 0; i < N; i++) {
//            for (int j = 0; j < M; j++) {
//                System.out.print(vis);
//            }
//        }
    }


    static class tuple {
        int move, x, y, cnt;

        public tuple(int move, int x, int y, int cnt) {
            this.move = move;
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }

    private static void init() throws IOException {
        st = new StringTokenizer(br.readLine());

        K = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());

        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());

        board = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int k = 0; k < M; k++) {
                board[i][k] = Integer.parseInt(st.nextToken());
            }
        }
    }
}
